database/engine/metadata_log/metadatalog.c

```diff
@@ -0,0 +1,342 @@
+{
+    return 0;
+}
+
+static void metalog_cleanup_finished_threads(struct metalog_worker_config *wc)
+{
+    if (unlikely(wc->cleanup_thread_compacting_files)) {
+        /* TODO: cleanup compaction */
+        return;
+    }
+}
+
+static void metalog_init_cmd_queue(struct metalog_worker_config *wc)
+{
+    wc->cmd_queue.head = wc->cmd_queue.tail = 0;
+    wc->queue_size = 0;
+    assert(0 == uv_cond_init(&wc->cmd_cond));
+    assert(0 == uv_mutex_init(&wc->cmd_mutex));
+}
+
+void metalog_enq_cmd(struct metalog_worker_config *wc, struct metalog_cmd *cmd)
+{
+    unsigned queue_size;
+
+    /* wait for free space in queue */
+    uv_mutex_lock(&wc->cmd_mutex);
+    while ((queue_size = wc->queue_size) == METALOG_CMD_Q_MAX_SIZE) {
+        uv_cond_wait(&wc->cmd_cond, &wc->cmd_mutex);
+    }
+    assert(queue_size < METALOG_CMD_Q_MAX_SIZE);
+    /* enqueue command */
+    wc->cmd_queue.cmd_array[wc->cmd_queue.tail] = *cmd;
+    wc->cmd_queue.tail = wc->cmd_queue.tail != METALOG_CMD_Q_MAX_SIZE - 1 ?
+                         wc->cmd_queue.tail + 1 : 0;
+    wc->queue_size = queue_size + 1;
+    uv_mutex_unlock(&wc->cmd_mutex);
+
+    /* wake up event loop */
+    assert(0 == uv_async_send(&wc->async));
+}
+
+struct metalog_cmd metalog_deq_cmd(struct metalog_worker_config *wc)
+{
+    struct metalog_cmd ret;
+    unsigned queue_size;
+
+    uv_mutex_lock(&wc->cmd_mutex);
+    queue_size = wc->queue_size;
+    if (queue_size == 0) {
+        ret.opcode = METALOG_NOOP;
+    } else {
+        /* dequeue command */
+        ret = wc->cmd_queue.cmd_array[wc->cmd_queue.head];
+        if (queue_size == 1) {
+            wc->cmd_queue.head = wc->cmd_queue.tail = 0;
+        } else {
+            wc->cmd_queue.head = wc->cmd_queue.head != RRDENG_CMD_Q_MAX_SIZE - 1 ?
+                                 wc->cmd_queue.head + 1 : 0;
+        }
+        wc->queue_size = queue_size - 1;
+
+        /* wake up producers */
+        uv_cond_signal(&wc->cmd_cond);
+    }
+    uv_mutex_unlock(&wc->cmd_mutex);
+
+    return ret;
+}
+
+static void async_cb(uv_async_t *handle)
+{
+    uv_stop(handle->loop);
+    uv_update_time(handle->loop);
+    debug(D_METADATALOG, "%s called, active=%d.", __func__, uv_is_active((uv_handle_t *)handle));
+}
+
+/* Flushes metadata log when timer expires */
+#define TIMER_PERIOD_MS (5000)
+
+static void timer_cb(uv_timer_t* handle)
+{
+    struct metalog_worker_config* wc = handle->data;
+
+    uv_stop(handle->loop);
+    uv_update_time(handle->loop);
+    metalog_test_quota(wc);
+    debug(D_METADATALOG, "%s: timeout reached.", __func__);
+#ifdef NETDATA_INTERNAL_CHECKS
+    {
+        char buf[4096];
+        debug(D_METADATALOG, "%s", get_metalog_statistics(wc->ctx, buf, sizeof(buf)));
+    }
+#endif
+    mlf_flush_records_buffer(wc);
+}
+
+#define MAX_CMD_BATCH_SIZE (256)
+
+void metalog_worker(void* arg)
+{
+    struct metalog_worker_config *wc = arg;
+    struct metalog_instance *ctx = wc->ctx;
+    uv_loop_t* loop;
```

Comment:
Why the heap instead of the stack?

Reply:
This is just following the best practices of `test/` code in libuv repo. There is no real reason behind this.